File: web/src/util/cost.ts
// Token cost helpers — the TypeScript half of the usage cost model. Rates
// are NEVER hardcoded here: they come from the `CostTable` the frontend
// fetches from `GET /api/usage/costs`, which the backend builds from its
// single source of truth in `src/routes/usage/cost.rs`. These helpers mirror
// the Rust `token_cost` / `usage_cost` so client-side trend math matches the
// server's `est_cost` exactly.

import type { CostTable, ModelRates } from '../types/api'

/** Token kind a rate applies to — mirrors the Rust `TokenKind` enum and the
 *  four billed columns on `usage_events`. */
export type TokenKind = 'input' | 'output' | 'cache_read' | 'cache_creation'

/** Usable context-window size (tokens) assumed when a row carries no model,
 *  or one we hold no window for. Every standard Claude tier exposes a 200K
 *  window, so it is the safe assumption — but callers must LABEL it as a
 *  default rather than presenting it as the model's real limit (see
 *  [`contextWindowInfo`]). Lives here, next to the rate table, so the limit
 *  is part of the shared cost/model module rather than hardcoded in a
 *  component. */
export const DEFAULT_CONTEXT_WINDOW = 200_000

/** Window of the long-context model aliases the registry advertises. Those
 *  ids are spelled with a `[1m]` suffix on every tier (`opus[1m]`,
 *  `claude-fable-5[1m]`), so the suffix — not an id list that churns with
 *  each model release — is what [`contextWindowInfo`] matches on. */
export const LONG_CONTEXT_WINDOW = 1_000_000

/** Per-model context-window overrides, keyed by bare model id. Standard
 *  Claude tiers share the 200K default and long-context tiers are matched
 *  by their `[1m]` suffix; Grok 4.5/4.6 advertise a 500K window. */
const CONTEXT_WINDOWS: Record<string, number> = {
  'grok-4.5': 500_000,
  'grok-4.6': 500_000,
  // Legacy CLI ids still stored on old sessions; same window as 4.5.
  'grok-build': 500_000,
  'grok-build-0.1': 500_000,
}

/** Normalize a model id to its bare form: usage rows and `sessions.model`
 *  may carry a `provider:` prefix and an `@account` suffix. Mirrors
 *  `bare_model_id` in `src/routes/usage/cost.rs`. */
export function bareModelId(model: string): string {
  const withoutAccount = model.split('@')[0]
  const colon = withoutAccount.indexOf(':')
  return colon > 0 ? withoutAccount.slice(colon + 1) : withoutAccount
}

/** The context window to measure a session's occupancy against, plus whether
 *  it is the model's actual window (`known: true`) or just
 *  [`DEFAULT_CONTEXT_WINDOW`] standing in for a model we can't resolve.
 *  Callers MUST surface that distinction: a 1M-context session measured
 *  against 200K renders a full, red gauge that is simply false, and an
 *  unlabelled 200K denominator claims a certainty we don't have.
 *
 *  Pass the fetched `CostTable` to recognize the standard 200K tiers: the
 *  table is keyed by the ids the running binary actually advertises (see the
 *  Rust `cost_table`), so membership in it — rather than a model list
 *  duplicated here and left to rot — is what makes a window known. */
export function contextWindowInfo(
  model: string | null | undefined,
  table?: CostTable,
): { limit: number; known: boolean } {
  if (model) {
    const bare = bareModelId(model)
    const override = CONTEXT_WINDOWS[bare]
    if (override) return { limit: override, known: true }
    if (bare.endsWith('[1m]')) return { limit: LONG_CONTEXT_WINDOW, known: true }
    if (table && (table.rates[bare] || table.rates[model])) {
      return { limit: DEFAULT_CONTEXT_WINDOW, known: true }
    }
  }
  return { limit: DEFAULT_CONTEXT_WINDOW, known: false }
}

/** Just the denominator from [`contextWindowInfo`], for callers that already
 *  show the model alongside it. */
export function contextWindowFor(model: string | null | undefined, table?: CostTable): number {
  return contextWindowInfo(model, table).limit
}

const RATE_FIELD: Record<TokenKind, keyof ModelRates> = {
  input: 'input_per_mtok',
  output: 'output_per_mtok',
  cache_read: 'cache_read_per_mtok',
  cache_creation: 'cache_creation_per_mtok',
}

/** Rates for a model id from a fetched `CostTable`, tolerating a `provider:`
 *  prefix and `@account` suffix usage rows may carry (via `bareModelId`).
 *  Falls back to the Opus tier (the backend's default) when the model is
 *  unknown, so an unrecognized model is never silently free. Returns null
 *  only for an empty table. */
export function ratesFor(table: CostTable, model: string | null): ModelRates | null {
  const rates = table.rates
  if (model) {
    const bare = bareModelId(model)
    if (rates[bare]) return rates[bare]
    if (rates[model]) return rates[model]
  }
  // The backend's default tier is Opus; match it when present, otherwise
  // the priciest known rate, so estimates stay conservative.
  if (rates['claude-opus-4-8']) return rates['claude-opus-4-8']
  let fallback: ModelRates | null = null
  for (const r of Object.values(rates)) {
    if (!fallback || r.output_per_mtok > fallback.output_per_mtok) fallback = r
  }
  return fallback
}

/** USD cost of `tokens` tokens of one `kind` for `model`, priced against a
 *  fetched `CostTable`. Mirrors the Rust `token_cost`. */
export function tokenCost(
  table: CostTable,
  model: string | null,
  kind: TokenKind,
  tokens: number,
): number {
  const rates = ratesFor(table, model)
  if (!rates) return 0
  return (tokens / 1_000_000) * rates[RATE_FIELD[kind]]
}

/** The four billed token slices of one usage record. */
export interface BilledTokens {
  input_tokens: number
  output_tokens: number
  cache_read_tokens: number
  cache_creation_tokens: number
}
/** Sum of the four billed slices — the figure the dashboard labels “Billed
 *  Tokens”. Deliberately NOT the provider-reported `total_tokens` roll-up:
 *  only this sum is guaranteed to reconcile with the Input / Output / Cache
 *  figures shown beside it and with the per-session rows, so every tokens
 *  total on the dashboard is computed from here. */
export function billedTokens(slices: BilledTokens): number {
  return (
    slices.input_tokens +
    slices.output_tokens +
    slices.cache_read_tokens +
    slices.cache_creation_tokens
  )
}

/** Total USD cost of one usage record's four billed slices. Mirrors the Rust
 *  `usage_cost`; the `total`/`context` roll-ups are intentionally not priced
 *  (they overlap the billed slices and would double-count). */
export function usageCost(table: CostTable, model: string | null, slices: BilledTokens): number {
  return (
    tokenCost(table, model, 'input', slices.input_tokens) +
    tokenCost(table, model, 'output', slices.output_tokens) +
    tokenCost(table, model, 'cache_read', slices.cache_read_tokens) +
    tokenCost(table, model, 'cache_creation', slices.cache_creation_tokens)
  )
}

File: src/routes/usage/cost.rs
//! Per-model token pricing — the single source of truth for usage cost.
//!
//! Rates are USD per million tokens, split by token kind, and the whole
//! usage dashboard prices tokens through [`token_cost`] / [`usage_cost`] so
//! rates live in exactly one place (this module). Backend aggregation calls
//! these helpers directly; the frontend fetches the serialized [`CostTable`]
//! from `GET /api/usage/costs` and prices client-side trends with the same
//! numbers — so Rust and TS never hardcode rates independently.
//!
//! When the Claude registry (`crate::provider::claude::discover_models`) or
//! the Grok seed (`crate::provider::grok::default_models`) gains or renames
//! a model, update [`rates_for`] here. The `every_registry_model_is_priced`
//! test fails if a registry model is left without an explicit tier.

use std::collections::BTreeMap;

use serde::Serialize;

/// Which slice of a turn's token usage a rate applies to. Mirrors the
/// per-kind columns on the `usage_events` table.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum TokenKind {
    Input,
    Output,
    CacheRead,
    CacheCreation,
}

/// USD-per-million-token rates for one model, by token kind. Serialized as a
/// value in the [`CostTable`] map the frontend fetches.
#[derive(Debug, Clone, Copy, PartialEq, Serialize)]
pub struct ModelRates {
    pub input_per_mtok: f64,
    pub output_per_mtok: f64,
    pub cache_read_per_mtok: f64,
    pub cache_creation_per_mtok: f64,
}

impl ModelRates {
    /// The rate (USD per million tokens) for a single token kind.
    pub fn rate(&self, kind: TokenKind) -> f64 {
        match kind {
            TokenKind::Input => self.input_per_mtok,
            TokenKind::Output => self.output_per_mtok,
            TokenKind::CacheRead => self.cache_read_per_mtok,
            TokenKind::CacheCreation => self.cache_creation_per_mtok,
        }
    }
}

// Published rate tiers (USD per million tokens). Cache-read is the cheap
// hit; cache-creation (cache write) carries a premium over base input.
const OPUS: ModelRates = ModelRates {
    input_per_mtok: 15.0,
    output_per_mtok: 75.0,
    cache_read_per_mtok: 1.5,
    cache_creation_per_mtok: 18.75,
};
const SONNET: ModelRates = ModelRates {
    input_per_mtok: 3.0,
    output_per_mtok: 15.0,
    cache_read_per_mtok: 0.3,
    cache_creation_per_mtok: 3.75,
};
const HAIKU: ModelRates = ModelRates {
    input_per_mtok: 0.8,
    output_per_mtok: 4.0,
    cache_read_per_mtok: 0.08,
    cache_creation_per_mtok: 1.0,
};

// xAI published short-context rates (prompt < 200k). Cache-creation has
// no published write premium, so it matches input. Long-context (≥200k)
// doubles these; we price the common short-context tier.
const GROK_46: ModelRates = ModelRates {
    input_per_mtok: 2.0,
    output_per_mtok: 6.0,
    cache_read_per_mtok: 0.50,
    cache_creation_per_mtok: 2.0,
};
const GROK_45: ModelRates = ModelRates {
    input_per_mtok: 2.0,
    output_per_mtok: 6.0,
    cache_read_per_mtok: 0.30,
    cache_creation_per_mtok: 2.0,
};

/// Fallback for an unrecognized model id (a renamed/removed registry model
/// or a non-Claude provider). Priced at the Opus tier so an unknown model
/// is never silently free — the most expensive tier keeps cost estimates
/// conservative rather than misleadingly low.
const DEFAULT_RATES: ModelRates = OPUS;

/// Strip a `provider:` prefix and `@account` suffix so rate/window lookups
/// key on the bare model id. `claude-opus-4-8` and `grok-4.5` have no
/// colon and pass through; `grok:grok-4.5@acc` becomes `grok-4.5`.
fn bare_model_id(model: &str) -> &str {
    let (model, _acct) = crate::provider::registry::split_model_account(model);
    match model.split_once(':') {
        Some((_, rest)) if !rest.is_empty() => rest,
        _ => model,
    }
}

/// Exact published rates for a known model id, or `None` for a model
/// we carry no price for. Accepts ids with or without a provider
/// prefix and with or without an `@account` suffix.
pub fn known_rates_for(model: &str) -> Option<ModelRates> {
    let id = bare_model_id(model);
    match id {
        "claude-opus-5" | "claude-opus-4-8" | "claude-opus-4-7" | "claude-opus-4-6"
        | "claude-fable-5" => Some(OPUS),
        "claude-sonnet-5" | "claude-sonnet-4-6" => Some(SONNET),
        "claude-haiku-4-5" => Some(HAIKU),
        "grok-4.6" => Some(GROK_46),
        "grok-4.5" | "grok-build" | "grok-build-0.1" => Some(GROK_45),
        _ => None,
    }
}

/// Rates for a model id. Accepts ids with or without a provider prefix
/// (usage rows store either form — see `usage_events.model`). Unknown
/// models fall back to the conservative default tier.
pub fn rates_for(model: Option<&str>) -> ModelRates {
    match model {
        Some(m) => known_rates_for(m).unwrap_or(DEFAULT_RATES),
        None => DEFAULT_RATES,
    }
}

/// Cost in USD of `tokens` tokens of a single `kind` for `model`.
pub fn token_cost(model: Option<&str>, kind: TokenKind, tokens: i64) -> f64 {
    tokens as f64 / 1_000_000.0 * rates_for(model).rate(kind)
}

/// Total USD cost of one usage record's four token slices. `total_tokens`
/// and `context_tokens` are intentionally NOT priced here — they are
/// roll-ups/snapshots that overlap the four billed slices, so pricing them
/// would double-count.
pub fn usage_cost(
    model: Option<&str>,
    input_tokens: i64,
    output_tokens: i64,
    cache_read_tokens: i64,
    cache_creation_tokens: i64,
) -> f64 {
    token_cost(model, TokenKind::Input, input_tokens)
        + token_cost(model, TokenKind::Output, output_tokens)
        + token_cost(model, TokenKind::CacheRead, cache_read_tokens)
        + token_cost(model, TokenKind::CacheCreation, cache_creation_tokens)
}

/// The full per-model rate table the frontend fetches once at boot. Keyed by
/// bare model id; values are the same rates [`token_cost`] applies.
#[derive(Debug, Clone, Serialize)]
pub struct CostTable {
    pub rates: BTreeMap<String, ModelRates>,
}

/// Build the rate table from the model registry so every advertised model
/// carries a published rate. The frontend caches this and prices its own
/// trend lines with it, matching the backend's `est_cost` exactly.
pub fn cost_table() -> CostTable {
    let mut rates = BTreeMap::new();
    for m in crate::provider::claude::discover_models() {
        rates.insert(m.id.clone(), rates_for(Some(&m.id)));
    }
    for m in crate::provider::grok::default_models() {
        rates.insert(m.id.clone(), rates_for(Some(&m.id)));
    }
    CostTable { rates }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn approx(a: f64, b: f64) {
        assert!((a - b).abs() < 1e-9, "expected {b}, got {a}");
    }

    #[test]
    fn token_cost_matches_known_rate_times_tokens() {
        // 1M output tokens on Opus = $75.00.
        approx(
            token_cost(Some("claude-opus-4-8"), TokenKind::Output, 1_000_000),
            75.0,
        );
        // 2M input tokens on Opus = 2 * $15.00 = $30.00.
        approx(
            token_cost(Some("claude-opus-4-8"), TokenKind::Input, 2_000_000),
            30.0,
        );
        // The `claude:` provider prefix is stripped before lookup.
        approx(
            token_cost(
                Some("claude:claude-sonnet-4-6"),
                TokenKind::Output,
                1_000_000,
            ),
            15.0,
        );
        // Grok is priced at xAI's published short-context rates, not Opus.
        approx(
            token_cost(Some("grok:grok-4.5"), TokenKind::Output, 1_000_000),
            6.0,
        );
        approx(
            token_cost(Some("grok-4.6@acc_1"), TokenKind::CacheRead, 1_000_000),
            0.50,
        );
        // 500k cache-read tokens on Haiku = 0.5 * $0.08 = $0.04.
        approx(
            token_cost(Some("claude-haiku-4-5"), TokenKind::CacheRead, 500_000),
            0.04,
        );

        // A full record sums its four billed slices and ignores the
        // total/context roll-ups.
        let total = usage_cost(Some("claude-opus-4-8"), 1_000_000, 1_000_000, 0, 0);
        approx(total, 90.0);

        // Unknown / missing model falls back to the default (Opus) tier
        // rather than pricing at zero.
        approx(
            token_cost(Some("some-future-model"), TokenKind::Output, 1_000_000),
            token_cost(Some("claude-opus-4-8"), TokenKind::Output, 1_000_000),
        );
        approx(token_cost(None, TokenKind::Output, 1_000_000), 75.0);
    }

    #[test]
    fn every_registry_model_is_priced() {
        let table = cost_table();
        for model in crate::provider::claude::discover_models()
            .into_iter()
            .chain(crate::provider::grok::default_models())
        {
            let rates = table
                .rates
                .get(&model.id)
                .unwrap_or_else(|| panic!("no rate for registry model {}", model.id));
            assert!(
                rates.input_per_mtok > 0.0 && rates.output_per_mtok > 0.0,
                "registry model {} priced at zero",
                model.id
            );
        }
        assert_eq!(table.rates["grok-4.5"].output_per_mtok, 6.0);
        assert_eq!(table.rates["grok-4.6"].cache_read_per_mtok, 0.50);
    }
}
